# Import thư viện và tải cần thiết     

In [1]:
pip install -U scikit-learn 


Note: you may need to restart the kernel to use updated packages.


In [2]:
import pandas as pd
import os
from sklearn.model_selection import train_test_split

# Đọc dữ liệu

In [47]:
# Đọc dữ liệu bằng pandas
df = pd.read_csv('titanic_train.csv')



In [48]:
# Thông tin các cột của bộ data
df.columns


Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [49]:
# Xem type data của các cột
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


# Tiền xử lý
Có một số dữ liệu ta sẽ không cho vào trong model để train như Name vì nó không thể hiện được tính logic của bài toán ( dựa vào tên không thể biết được bạn chết hay sống)
Có một số dữ liệu như Sex, Embarked(điểm đến) không phải dạng số nên ta phải chuyển nó về dạng số để bỏ vào model

In [6]:
# Chuyển dữ liệu
convert = pd.get_dummies(df.Embarked, prefix = 'Embarked')
convert


Embarked_C  Embarked_Q  Embarked_S
0             0           0           1
1             1           0           0
2             0           0           1
3             0           0           1
4             0           0           1
..          ...         ...         ...
886           0           0           1
887           0           0           1
888           0           0           1
889           1           0           0
890           0           1           0

[891 rows x 3 columns]

In [7]:
# Xóa cột cũ
df.drop(["Embarked"], axis = 1, inplace = True)


In [8]:
# Thêm vào dataset cột vừa xử lý
df = df.join(convert)
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin  Embarked_C  Embarked_Q  \
0        0         A/5 21171   7.2500   NaN           0           0   
1        0          PC 17599  71.2833   C85           1           0   
2        0  STON/O2. 3101282   7.9250   NaN           0           0   
3        0            113803  53.1000  C123           0           0   
4        0            373450   8.0500   NaN           0           0   
..     ...               ...      ...   ...         ...         ...   
886      0            211536  13.0000   NaN           0           0   
887      0            112053  30.0000   B42           0           0   
888      2        W./C. 6607  23.4500   NaN           0           0   
889      0            111369  30.0000  C148           1           0   
890      0            370376   7.7500   NaN           0           1   

     Embarked_S  
0             1  
1             0  
2             1  
3             1  
4             1  
..          ...  
886           1  
887           1  
888           1  
889           0  
890           0  

[891 rows x 14 columns]

In [9]:
df.Sex = df.Sex.map({'male': 0, 'female': 1})  # có thể dùng pd.get_dummies() như cột trên nhưng do chỉ có 2 giá trị nên ta set bằng pd.map() lun cho nhanh
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name  Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    0  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...    1  38.0      1   
2                               Heikkinen, Miss. Laina    1  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)    1  35.0      1   
4                             Allen, Mr. William Henry    0  35.0      0   
..                                                 ...  ...   ...    ...   
886                              Montvila, Rev. Juozas    0  27.0      0   
887                       Graham, Miss. Margaret Edith    1  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"    1   NaN      1   
889                              Behr, Mr. Karl Howell    0  26.0      0   
890                                Dooley, Mr. Patrick    0  32.0      0   

     Parch            Ticket     Fare Cabin  Embarked_C  Embarked_Q  \
0        0         A/5 21171   7.2500   NaN           0           0   
1        0          PC 17599  71.2833   C85           1           0   
2        0  STON/O2. 3101282   7.9250   NaN           0           0   
3        0            113803  53.1000  C123           0           0   
4        0            373450   8.0500   NaN           0           0   
..     ...               ...      ...   ...         ...         ...   
886      0            211536  13.0000   NaN           0           0   
887      0            112053  30.0000   B42           0           0   
888      2        W./C. 6607  23.4500   NaN           0           0   
889      0            111369  30.0000  C148           1           0   
890      0            370376   7.7500   NaN           0           1   

     Embarked_S  
0             1  
1             0  
2             1  
3             1  
4             1  
..          ...  
886           1  
887           1  
888           1  
889           0  
890           0  

[891 rows x 14 columns]

In [10]:
x = df.drop(['Survived'], axis = 1)
y = df.Survived.copy()

In [11]:
# Những biến không thể hiện tính logic thì xóa đi
x.drop(['Cabin'],axis = 1,inplace = True)
x.drop(['Ticket'],axis = 1,inplace = True)
x.drop(['Name'],axis = 1,inplace = True)
x.drop(['PassengerId'],axis = 1,inplace = True)

In [12]:
x.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Pclass      891 non-null    int64  
 1   Sex         891 non-null    int64  
 2   Age         714 non-null    float64
 3   SibSp       891 non-null    int64  
 4   Parch       891 non-null    int64  
 5   Fare        891 non-null    float64
 6   Embarked_C  891 non-null    uint8  
 7   Embarked_Q  891 non-null    uint8  
 8   Embarked_S  891 non-null    uint8  
dtypes: float64(2), int64(4), uint8(3)
memory usage: 44.5 KB


In [13]:
# Trong thực tế có một số chỗ mà chưa được điền dữ liệu ( missing value) , cách đơn giản nhất để xử lý việc này mà không làm mất số lượng dữ liệu là thay nó bằng giá trị tủng bình
x.isnull().values.any() # Check thử xem có missing value không
x[pd.isnull(x).any(axis = 1)] # Xem những hàng có missing value

Pclass  Sex  Age  SibSp  Parch     Fare  Embarked_C  Embarked_Q  \
5         3    0  NaN      0      0   8.4583           0           1   
17        2    0  NaN      0      0  13.0000           0           0   
19        3    1  NaN      0      0   7.2250           1           0   
26        3    0  NaN      0      0   7.2250           1           0   
28        3    1  NaN      0      0   7.8792           0           1   
..      ...  ...  ...    ...    ...      ...         ...         ...   
859       3    0  NaN      0      0   7.2292           1           0   
863       3    1  NaN      8      2  69.5500           0           0   
868       3    0  NaN      0      0   9.5000           0           0   
878       3    0  NaN      0      0   7.8958           0           0   
888       3    1  NaN      1      2  23.4500           0           0   

     Embarked_S  
5             0  
17            1  
19            0  
26            0  
28            0  
..          ...  
859           0  
863           1  
868           1  
878           1  
888           1  

[177 rows x 9 columns]

In [14]:
# Vậy là , ta tiến hành xử lý
x.Age.fillna(x.Age.mean(),inplace = True)

In [15]:
x.isnull().values.any() # Check lại


False

# Bắt đầu train

In [16]:
from sklearn.linear_model import LogisticRegression # sử dụng pakkage hồi quy logictic

In [30]:
# 80% train, 20% test
x_train, x_valid,y_train,y_valid = train_test_split(x,y,test_size = 0.2,random_state = 7)


725    0
861    0
528    0
46     0
627    1
      ..
809    1
305    0
538    0
825    0
874    1
Name: Sex, Length: 179, dtype: int64

In [18]:
model = LogisticRegression() 

In [19]:
model.fit(x_train,y_train) # fit model

c:\Users\ADMIN\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [20]:
model.score(x_train,y_train ) # đánh gái mô hình

0.8089887640449438

In [21]:
b = model.intercept_ # bias
b = b[0]
b

1.7195131269030053

In [22]:
w = model.coef_ # hệ số weight[i]
w = w[0]
w

array([-1.00636829e+00,  2.79575384e+00, -4.29590340e-02, -3.69093008e-01,
        2.00233492e-03,  1.01303408e-03,  8.31024800e-01,  3.85919699e-01,
        2.44480303e-01])

# Viết hàm dự đoán

#
Nhắc lại lý thuyết
Hàm dự đoán : f(z) = (1 + e ^ (- z)) ^ -1
    với z là hàm tuyến tính n biến
        Nếu f(z) >= 0.5 -> f(z) = 0 (True)
        else :             f(z) = 1 (False)

In [23]:
def sizmoid(x):
    e = 2.71828182846
    sum_linear = b
    for i in range(len(w)):
        sum_linear+= w[i] * x[i]
    return 1 / (1 + e ** (- sum_linear ))

In [24]:
def predict(x):
    if sizmoid(x) >= 0.5:
        return 0
    else:
        return 1

# Đi dự đoán
# 
PassengerId: 1
Name : Uzumaki Naruto
Pclass: 3 (train)
Sex : Male (train)
Age : 22 (train)
SibSp : 1 (train)
Parch : 0 (train)
Ticket : A/5 21171
Fare: 7.25 (train)
Cabin: C85 
Embarked : S (train)

Theo test thì thằng ninja làng lá này 'sống'

In [25]:
pre = [1,3,0,1,0,7.25,0,0,1]

In [26]:
if predict(pre) == 0:
    print("Sống")
else:
    print('Chết cmnr')

Sống


# Báo Cáo

In [69]:
x_sub = x_valid.index.array
x_sub

<PandasArray>
[725, 861, 528,  46, 627, 357, 890, 179, 562, 879,
 ...
 326, 776, 740, 261,  22, 809, 305, 538, 825, 874]
Length: 179, dtype: int64

In [77]:
x_valid

Pclass  Sex        Age  SibSp  Parch      Fare  Embarked_C  Embarked_Q  \
725       3    0  20.000000      0      0    8.6625           0           0   
861       2    0  21.000000      1      0   11.5000           0           0   
528       3    0  39.000000      0      0    7.9250           0           0   
46        3    0  29.699118      1      0   15.5000           0           1   
627       1    1  21.000000      0      0   77.9583           0           0   
..      ...  ...        ...    ...    ...       ...         ...         ...   
809       1    1  33.000000      1      0   53.1000           0           0   
305       1    0   0.920000      1      2  151.5500           0           0   
538       3    0  29.699118      0      0   14.5000           0           0   
825       3    0  29.699118      0      0    6.9500           0           1   
874       2    1  28.000000      1      0   24.0000           1           0   

     Embarked_S  
725           1  
861           1  
528           1  
46            0  
627           1  
..          ...  
809           1  
305           1  
538           1  
825           0  
874           0  

[179 rows x 9 columns]

In [78]:
li_main = []
li = []
for i in x_valid.index:
    li = []
    for j in x_valid.columns:
        li.append(x_valid[j][i])
    li_main.append(li)

li_main
        



[[3, 0, 20.0, 0, 0, 8.6625, 0, 0, 1],
 [2, 0, 21.0, 1, 0, 11.5, 0, 0, 1],
 [3, 0, 39.0, 0, 0, 7.925, 0, 0, 1],
 [3, 0, 29.69911764705882, 1, 0, 15.5, 0, 1, 0],
 [1, 1, 21.0, 0, 0, 77.9583, 0, 0, 1],
 [2, 1, 38.0, 0, 0, 13.0, 0, 0, 1],
 [3, 0, 32.0, 0, 0, 7.75, 0, 1, 0],
 [3, 0, 36.0, 0, 0, 0.0, 0, 0, 1],
 [2, 0, 28.0, 0, 0, 13.5, 0, 0, 1],
 [1, 1, 56.0, 0, 1, 83.1583, 1, 0, 0],
 [2, 1, 29.0, 1, 0, 26.0, 0, 0, 1],
 [3, 0, 29.69911764705882, 8, 2, 69.55, 0, 0, 1],
 [2, 1, 34.0, 0, 1, 23.0, 0, 0, 1],
 [2, 1, 8.0, 0, 2, 26.25, 0, 0, 1],
 [3, 0, 30.5, 0, 0, 8.05, 0, 0, 1],
 [3, 0, 29.69911764705882, 0, 0, 7.7375, 0, 1, 0],
 [3, 0, 22.0, 0, 0, 7.2292, 1, 0, 0],
 [2, 0, 29.69911764705882, 0, 0, 0.0, 0, 0, 1],
 [3, 0, 34.0, 1, 1, 14.4, 0, 0, 1],
 [2, 0, 18.0, 0, 0, 13.0, 0, 0, 1],
 [1, 0, 27.0, 0, 0, 76.7292, 1, 0, 0],
 [3, 0, 21.0, 0, 0, 8.6625, 0, 0, 1],
 [1, 0, 80.0, 0, 0, 30.0, 0, 0, 1],
 [3, 0, 12.0, 1, 0, 11.2417, 1, 0, 0],
 [3, 0, 44.0, 0, 0, 8.05, 0, 0, 1],
 [3, 0, 9.0, 1, 1, 15.9, 0, 

In [83]:
y_pre = []
for i in li_main:
    y_pre.append(predict(i))
y_pre

[1,
 1,
 1,
 1,
 0,
 0,
 1,
 1,
 1,
 0,
 0,
 1,
 0,
 0,
 1,
 1,
 1,
 1,
 1,
 1,
 0,
 1,
 1,
 1,
 1,
 1,
 0,
 1,
 1,
 0,
 0,
 0,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 0,
 0,
 1,
 1,
 1,
 1,
 1,
 1,
 0,
 0,
 1,
 0,
 1,
 0,
 1,
 1,
 0,
 0,
 0,
 1,
 1,
 1,
 0,
 1,
 1,
 0,
 1,
 0,
 1,
 0,
 1,
 1,
 0,
 0,
 1,
 1,
 0,
 0,
 1,
 0,
 1,
 0,
 0,
 1,
 0,
 0,
 1,
 1,
 1,
 0,
 0,
 1,
 1,
 1,
 1,
 1,
 0,
 1,
 1,
 1,
 1,
 0,
 0,
 1,
 1,
 1,
 0,
 0,
 1,
 0,
 1,
 1,
 1,
 1,
 1,
 0,
 1,
 1,
 0,
 1,
 1,
 0,
 0,
 1,
 0,
 0,
 1,
 0,
 1,
 0,
 0,
 0,
 0,
 1,
 1,
 1,
 1,
 0,
 1,
 1,
 1,
 0,
 0,
 1,
 0,
 1,
 1,
 1,
 1,
 0,
 1,
 0,
 0,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 0,
 0,
 0,
 1,
 1,
 0]

In [86]:
from sklearn.metrics import classification_report
print(classification_report(y_valid, y_pre, target_names=['Alive', 'Dead']))

              precision    recall  f1-score   support

       Alive       0.30      0.17      0.22       110
        Dead       0.22      0.36      0.27        69

    accuracy                           0.25       179
   macro avg       0.26      0.27      0.24       179
weighted avg       0.27      0.25      0.24       179

